contracts/proxy/transparent/TransparentUpgradeableProxy.sol

```diff
@@ -46,12 +44,21 @@ interface ITransparentUpgradeableProxy is IERC1967 {
  * fully implement transparency without decoding reverts caused by selector clashes between the proxy and the
  * implementation.
  *
+ * IMPORTANT: This contract avoids unnecessary storage reads by setting the admin only during construction as an immutable variable,
```

Comment:
The docs for the contract need to be fully reviewed beyond this. There is a mention of "changing the admin above". We should emphasize that the admin is immutable, and the recommendation to use ProxyAdmin should now be even stronger, because it will be the only way to later change it.

Honestly I would even consider enforcing this via `extcodesize` but I know this will be pretty controversial. Cc: @Amxx

Reply:
> The docs for the contract need to be fully reviewed beyond this. There is a mention of "changing the admin" above.

I updated the previous part and I think there are no other places to update after checking across the codebase.

> We should emphasize that the admin is immutable, and the recommendation to use ProxyAdmin should now be even stronger, because it will be the only way to later change it.

Added a comment but you may want to review or propose a different layout. I don't want to extend the text much because I already think it's long.

> Honestly I would even consider enforcing this via extcodesize but I know this will be pretty controversial. Cc: @Amxx

I was thinking the same but I decided to avoid it because I don't think it changes anything:

- There's no way to use the `ERC1967Utils.changeAdmin(admin_)` function twice in the constructor within the same contract
- Even if there was a way, the implementation can always brick the contract by setting the admin slot.

Now I wonder if we might need a way to see if a Proxy is "healthy", meaning that the immutable address matches the admin storage slot. That could be done if we make `public` the `_admin` variable. Obviously with a different name or a getter.

Reply:
> Even if there was a way, the implementation can always brick the contract by setting the admin slot.

If the implementation changes the admin slot the contract is not bricked. You "just" need to figure out who the initial admin was (looking at the event emitted in the constructor) and use that. UIs may be broken, but the contract is not.

My understanding was, since its impossible to change, its unsafe to use an EOA for that, so we may want to enforce that the address corresponds to a contract already deployed, hoppinh it is a ProxyAdmin (or a mulsitig?).

At that point I'd just go for deploying our own ProxyAdmin.

Reply:
> That could be done if we make public the _admin variable. Obviously with a different name or a getter.

If you make anything public in the TransparentProxyAdmin, you are basically breaking the transparency. The idea is that ANY function call (that is not done by the admin) would be forwarded.

Reply:
I'm not sure I get what the overall idea is. We can't expose any getter to anyone but to the admin. Adding a getting that returns the admin, but that only the admin can call doesn't really work for me. 

AFAIK, the only absolutely reliable and immutable way to get the admin is to look at event history.

As for the "contract is bricked", I think its a question of what do you mean by that. I personnaly think of a "bricked contract" as something that you absolutely cannot recover from. Here the "standard workflow" is (at least temporarily) broken, but there exist a (more complex) path toward recovery.

Reply:
> I'm not sure I get what the overall idea is. We can't expose any getter to anyone but to the admin. Adding a getting that returns the admin, but that only the admin can call doesn't really work for me.
 
The overall idea is pretty clear I'd say:

> > Now I wonder if we might need a way to see if a Proxy is "healthy", meaning that the immutable address matches the admin storage slot.

> AFAIK, the only absolutely reliable and immutable way to get the admin is to look at event history.

This is expensive for DApps and an edge case. Taking Defender as an example, it already takes time in the UI to know that the contract features are (including Upgradeability), I think it's inconvenient adding a required step such as reading the events story.

> I think it's a question of what you mean by that.

I don't think it matters, we both agree it's a problem, and that's all I meant.

Reply:
Yes it's important to understand changing the slot does not brick the proxy. It's a relatively minor issue.

Reply:
That's understood since the beginning, but I'm worried the conversation is about the use of "brick" instead of an actual review.